src/internal/events/events.go

```diff
@@ -159,3 +162,88 @@ func tenantHash(tenID string) string {
 	sum := md5.Sum([]byte(tenID))
 	return fmt.Sprintf("%x", sum)
 }
+
+// ---------------------------------------------------------------------------
+// hashicorp PoC
+// ---------------------------------------------------------------------------
+
+type m string
+
+// metrics collection bucket
+const APICall m = "api_call"
+
+type sinkKey string
+
+const sinkCtxKey sinkKey = "corsoMetricsSink"
+
+// configurations
+const (
+	reportInterval    = 1 * time.Minute
+	retentionDuration = 2 * time.Minute
+)
+
+// NewMetrics embeds a metrics bus into the provided context.  The bus can be
+// utilized with calls like Inc and Since.
+func NewMetrics(ctx context.Context, w io.Writer) (context.Context, func()) {
+	var (
+		sink = metrics.NewInmemSink(reportInterval, retentionDuration)
+		cfg  = metrics.DefaultConfig("corso")
+		sig  = metrics.NewInmemSignal(sink, metrics.DefaultSignal, w)
+	)
+
+	gm, err := metrics.NewGlobal(cfg, sink)
+	if err != nil {
+		logger.CtxErr(ctx, err).Error("metrics bus constructor")
+		return ctx, func() { sig.Stop() }
+	}
+
+	stop := make(chan struct{})
+	go dumpMetrics(ctx, stop, sig)
+
+	flush := func() {
+		close(stop)
+		sig.Stop()
+		gm.Shutdown()
+	}
+
+	return context.WithValue(ctx, sinkCtxKey, sink), flush
+}
+
+func dumpMetrics(ctx context.Context, stop <-chan struct{}, sig *metrics.InmemSignal) {
```

Comment:
So if I understand this correctly - it's sending a signal on reportInterval - which triggers the output you provided?

Why is `NewInmemSink` configured with a `reportInterval` if we need to implement this ourselves? Also - is there a way to control the output? The GC stats are useful but I don't think we want them by default

Reply:
> So if I understand this correctly - it's sending a signal on reportInterval - which triggers the output you provided?

Correct.

> Why is NewInmemSink configured with a reportInterval if we need to implement this ourselves?

Ah, good question.  The sink interval doesn't have any relationship with the dump interval.  The sink interval's purpose is to bucket statistics to a given time-frame.  For example, if you want to get per-minute metrics, but only want to publish findings every 10 minutes, the you'd construct a sink with `inmemSink(1_minute, at_least_10_minutes)`, and time the log dump for once every 10 minutes.  That way, every log report contains 10 stacks of the stats block: one for each interval minute.

> Also - is there a way to control the output?

I'll look again, but I don't recall seeing any controls there.